File: android/app/src/main/kotlin/com/transport_system/ts_driver/pusher/channels/call_channel/channel/CallChannel.kt
package com.transport_system.ts_driver.pusher.channels.call_channel.channel

import android.content.Context
import com.pusher.client.channel.PrivateChannel
import com.pusher.client.channel.PrivateChannelEventListener
import com.pusher.client.channel.PusherEvent
import com.transport_system.ts_driver.data_providers.MyDetails
import com.transport_system.ts_driver.helpers.AppLogger
import com.transport_system.ts_driver.pusher.manager.PusherManager

class CallChannel {

    private var pusherManager = PusherManager.instance

    private var callChannel : PrivateChannel? = null


    companion object {
        val instance = CallChannel()
    }

    fun initialize(context: Context){
       try {
           if(callChannel?.isSubscribed == true){
               return
           }
           createAndSubscribeCallChannel(context = context)
       }
       catch (e:Exception){
           AppLogger.log("Exception while initializing the call channel ===> ${e.message}")
       }
    }

    fun getCallChannel(context: Context) : PrivateChannel? {
        try {
            pusherManager.ensureConnection(context = context)
            if(callChannel?.isSubscribed == false){
                createAndSubscribeCallChannel(context = context)
            }
            else{
                AppLogger.log("Call channel is subscribed while getting call channel.")
            }
        }
        catch (e:Exception){
            AppLogger.log("Exception while getting call channel ===> ${e.message}")
        }
        return callChannel
    }

    private fun createAndSubscribeCallChannel(context: Context) {
        try {
            val myDetails = MyDetails.loadFromSharedPrefs(context = context) ?: return
            val modelType = myDetails.modelType ?: return
            val userId = myDetails.applicantId ?: return

            pusherManager.ensureConnection(context = context)

            if(callChannel?.isSubscribed == true){
                AppLogger.log("Call Channel already subscribed")
                return
            }

            callChannel = pusherManager.getPusher(context = context)?.subscribePrivate("private-call-receiver-$modelType-$userId",object : PrivateChannelEventListener{
                override fun onEvent(event: PusherEvent?) {
                    AppLogger.log("Private call channel event received ===> event.data : ${event?.data}")
                }

                override fun onSubscriptionSucceeded(channelName: String?) {
                    AppLogger.log("Private call channel subscription succeeded ===> channelName : $channelName")
                }

                override fun onAuthenticationFailure(message: String?, e: java.lang.Exception?) {
                    AppLogger.log("Private call channel authentication failed ===> message : $message, exception : ${e?.message}")
                }
            })
        }
        catch (e:Exception){
            AppLogger.log("Exception while creating and subscribing call channel ===> ${e.message}")
        }
    }


}

File: ios/Runner/Pusher/Channels/Call Channel/CallChannel.swift
//
//  CallChannel.swift
//  Runner
//
//  Created by Hashim Khan on 24/01/2024.
//


import PusherSwift

class CallChannel {
    
    static let shared = CallChannel()
    
    
    private let pusherManager: PusherManager = PusherManager.shared
    
    
    // MARK: - Call Channle
    private var callChannel : PusherChannel?
    private let decoder = JSONDecoder()

    
    init() {
        createAndSubscribeCallChannel()
    }
    
    //
    // function to ensure socket connection and callChannel and return call channel
    func getCallChannel() -> PusherChannel? {
        pusherManager.ensureConnection()
        if callChannel == nil{
            createAndSubscribeCallChannel()
        }
        return callChannel
    }

    //
    // function that will ensure call channel subscriptin and bind events
    private func createAndSubscribeCallChannel() {
        
        guard let myDetails = MyDetails.loadFromSharedPrefs(),
              let modelType = myDetails.modelType,
              let userId = myDetails.applicantId else {
            return
        }
        
        pusherManager.ensureConnection()
        
        if callChannel?.subscribed == true {
            printLog("Call Channel already subscribed")
            return
        }
        
        callChannel = pusherManager.getPusher().subscribe(channelName: "private-call-receiver-\(modelType)-\(userId)")
    }
  
    func printLog(_ message: String) {
        print("Pusher Call Channel: \(message)")
    }
}



